sonar-xml-plugin/src/main/java/org/sonar/plugins/xml/checks/TabCharacterCheck.java

```diff
@@ -20,46 +20,61 @@
 package org.sonar.plugins.xml.checks;
 
 import java.util.ArrayList;
-import java.util.Collections;
 import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 import org.sonar.check.Rule;
-import org.sonar.check.RuleProperty;
 import org.sonar.plugins.xml.Utils;
+import org.sonar.plugins.xml.Xml;
+import org.sonarsource.analyzer.commons.annotations.DeprecatedRuleKey;
 import org.sonarsource.analyzer.commons.xml.XmlFile;
+import org.sonarsource.analyzer.commons.xml.XmlTextRange;
 import org.sonarsource.analyzer.commons.xml.checks.SonarXmlCheck;
 
 /**
  * RSPEC-105
  */
 @Rule(key = TabCharacterCheck.RULE_KEY)
+@DeprecatedRuleKey(ruleKey = "IllegalTabCheck", repositoryKey = Xml.REPOSITORY_KEY)
 public class TabCharacterCheck extends SonarXmlCheck {
 
-  public static final String RULE_KEY = "IllegalTabCheck";
-
-  @RuleProperty(key = "markAll", description = "Mark all tab errors", defaultValue = "false", type = "BOOLEAN")
-  private boolean markAll;
+  public static final String RULE_KEY = "S105";
+  private static final Pattern TABS_REGEX = Pattern.compile("\t++");
+  private static final int MAX_REPORTED_LOCATION = 21;
 
   @Override
   public void scanFile(XmlFile file) {
-    int lineNumber = 1;
-    List<Integer> secondaries = new ArrayList<>();
-
-    for (String line : Utils.splitLines(file.getContents())) {
-      if (line.indexOf('\t') != -1) {
-        secondaries.add(lineNumber);
-        if (!markAll) {
-          break;
+    if (file.getContents().indexOf('\t') == -1) {
+      return;
+    }
+    String[] lines = Utils.splitLines(file.getContents());
+    List<XmlTextRange> firstTabLocations = new ArrayList<>();
+    int extraTabsCount = 0;
+    for (int lineNumber = 1; lineNumber <= lines.length; lineNumber++) {
+      String line = lines[lineNumber - 1];
+      Matcher tabsMatcher = TABS_REGEX.matcher(line);
+      while (tabsMatcher.find()) {
+        if (firstTabLocations.size() < MAX_REPORTED_LOCATION) {
+          firstTabLocations.add(new XmlTextRange(lineNumber, tabsMatcher.start(), lineNumber, tabsMatcher.end()));
+        } else {
+          extraTabsCount += tabsMatcher.end() - tabsMatcher.start();
         }
       }
-      lineNumber++;
-    }
-    if (!secondaries.isEmpty()) {
-      reportIssueOnFile("Replace all tab characters in this file by sequences of white-spaces.", markAll ? secondaries : Collections.emptyList());
     }
+    reportIssue(firstTabLocations, extraTabsCount);
   }
 
-  public void setMarkAll(boolean markAll) {
-    this.markAll = markAll;
+  private void reportIssue(List<XmlTextRange> firstTabLocations, int extraTabsCount) {
+    XmlTextRange primaryLocation = firstTabLocations.get(0);
+    List<Secondary> secondaries = new ArrayList<>();
+    for (int i = 1; i < firstTabLocations.size(); i++) {
+      XmlTextRange range = firstTabLocations.get(i);
+      boolean hasSeveralTabs = range.getEndColumn() > range.getStartColumn() + 1;
```

Comment:
If it's only for a 's', I would simply use `Tab character(s)` in the message and drop the extra logic adding complexity for no reason. Also, make it start with capitalized `T`.